File: code/packages/dart/der-tlv/test/der_tlv_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:coding_adventures_der_tlv/der_tlv.dart';
import 'package:test/test.dart';

Map<String, dynamic> loadFixture() {
  final candidates = [
    '../../../../specs/fixtures/der-tlv-v1/cases.json',
    '../../../specs/fixtures/der-tlv-v1/cases.json',
  ];
  for (final candidate in candidates) {
    final file = File(candidate);
    if (file.existsSync()) {
      return jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
    }
  }
  throw StateError('unable to locate DER TLV fixture');
}

Uint8List materialize(List<dynamic> segments) {
  final builder = BytesBuilder(copy: false);
  for (final raw in segments) {
    final segment = raw as Map<String, dynamic>;
    if (segment['hex'] case final String value) {
      builder.add([
        for (var index = 0; index < value.length; index += 2)
          int.parse(value.substring(index, index + 2), radix: 16),
      ]);
    } else {
      builder.add(
        Uint8List(segment['count'] as int)
          ..fillRange(
            0,
            segment['count'] as int,
            int.parse(segment['repeat_hex'] as String, radix: 16),
          ),
      );
    }
  }
  return builder.takeBytes();
}

DerLimits limits(Map<String, dynamic> document, Map<String, dynamic> testCase) {
  final values = <String, dynamic>{
    ...document['defaults'] as Map<String, dynamic>,
    ...?testCase['limits'] as Map<String, dynamic>?,
  };
  return DerLimits(
    maxInputLen: values['max_input_len'] as int,
    maxValueLen: values['max_value_len'] == 'host-max'
        ? BigInt.from(0x7fffffffffffffff)
        : BigInt.from(values['max_value_len'] as int),
    maxElements: values['max_elements'] as int,
    maxTagNumber: values['max_tag_number'] as int,
  );
}

Map<String, dynamic> elementResult(DerElement element, int offset) => {
      'outcome': 'element',
      'element_offset': offset,
      'tag': {
        'class': element.tag.tagClass.wireName,
        'constructed': element.tag.constructed,
        'number': element.tag.number,
      },
      'header_len': element.header.length,
      'encoded_len': element.encoded.length,
      'remainder_offset': offset + element.encoded.length,
    };

Map<String, dynamic> errorResult(DerException error) => {
      'outcome': 'error',
      'error_id': error.kind.id,
      'offset': error.offset,
    };

Map<String, dynamic> runDecode(
  Map<String, dynamic> testCase,
  Uint8List input,
  DerLimits configured,
) {
  try {
    if (testCase['operation'] == 'decode-one') {
      final result = decodeOne(input, configured);
      final projection = elementResult(result.element, 0);
      expect(projection['remainder_offset'],
          input.length - result.remainder.length);
      return projection;
    }
    return elementResult(decodeExact(input, configured), 0);
  } on DerException catch (error) {
    return errorResult(error);
  }
}

Map<String, dynamic> runCursor(
  Map<String, dynamic> testCase,
  Uint8List input,
  DerLimits configured,
) {
  final cursor = DerCursor(input, configured);
  final events = <Map<String, dynamic>>[];
  for (final action in testCase['actions'] as List<dynamic>) {
    if (action == 'finish') {
      try {
        cursor.finish();
        events.add({'outcome': 'finished'});
      } on DerException catch (error) {
        events.add(errorResult(error));
      }
      continue;
    }
    final offset = input.length - cursor.remaining.length;
    try {
      final element = cursor.read();
      events.add(
        element == null ? {'outcome': 'end'} : elementResult(element, offset),
      );
    } on DerException catch (error) {
      events.add(errorResult(error));
    }
  }
  return {
    'events': events,
    'elements_read': cursor.elementsRead,
    'remaining_offset': input.length - cursor.remaining.length,
  };
}

void main() {
  final fixture = loadFixture();
  final cases = fixture['cases'] as List<dynamic>;

  test('pins the closed DER TLV profile', () {
    expect(cases, hasLength(54));
    expect(fixture['error_ids'] as List<dynamic>, hasLength(17));
  });

  for (final raw in cases) {
    final testCase = raw as Map<String, dynamic>;
    test(testCase['id'] as String, () {
      final input = materialize(testCase['input'] as List<dynamic>);
      final configured = limits(fixture, testCase);
      final actual = testCase['operation'] == 'cursor'
          ? runCursor(testCase, input, configured)
          : runDecode(testCase, input, configured);
      expect(actual, testCase['expected']);
      if (testCase['redacted_input_hex'] case final String hostile) {
        expect(jsonEncode(actual), isNot(contains(hostile)));
      }
    });
  }

  test('element views share the caller buffer', () {
    final input = Uint8List.fromList([0x04, 0x01, 0x2a]);
    final element = decodeExact(input);
    input[2] = 0x7f;
    expect(element.value, [0x7f]);
  });

  test('negative limits are rejected', () {
    expect(() => DerLimits(maxElements: -1), throwsArgumentError);
    expect(() => DerLimits(maxTagNumber: 0x100000000), throwsArgumentError);
  });
}

File: code/packages/typescript/der-tlv/tests/der-tlv.test.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { DerCursor, DerElement, DerError, DerLimits, decodeExact, decodeOne, defaultLimits } from "../src/index.js";

interface Segment { hex?: string; repeat_hex?: string; count?: number }
interface FixtureCase { id: string; operation: "decode-exact" | "decode-one" | "cursor"; input: Segment[]; limits?: Record<string, number | string>; actions?: ("read" | "finish")[]; redacted_input_hex?: string; expected: unknown }
interface Fixture { defaults: Record<string, number>; error_ids: string[]; cases: FixtureCase[] }

const path = fileURLToPath(new URL("../../../../specs/fixtures/der-tlv-v1/cases.json", import.meta.url));
const fixture = JSON.parse(readFileSync(path, "utf8")) as Fixture;

function bytes(segments: Segment[]): Uint8Array {
  const output: number[] = [];
  for (const segment of segments) {
    if (segment.hex !== undefined) {
      for (let index = 0; index < segment.hex.length; index += 2) output.push(Number.parseInt(segment.hex.slice(index, index + 2), 16));
    } else {
      output.push(...new Array(segment.count!).fill(Number.parseInt(segment.repeat_hex!, 16)) as number[]);
    }
  }
  return Uint8Array.from(output);
}

function configured(testCase: FixtureCase): DerLimits {
  const values: Record<string, number | string> = { ...fixture.defaults, ...testCase.limits };
  return {
    maxInputLen: values.max_input_len as number,
    maxValueLen: values.max_value_len === "host-max" ? BigInt(Number.MAX_SAFE_INTEGER) : BigInt(values.max_value_len as number),
    maxElements: values.max_elements as number,
    maxTagNumber: values.max_tag_number as number,
  };
}

function elementResult(element: DerElement, offset: number): unknown {
  return { outcome: "element", element_offset: offset, tag: element.tag, header_len: element.header.length, encoded_len: element.encoded.length, remainder_offset: offset + element.encoded.length };
}

function errorResult(error: unknown): unknown {
  expect(error).toBeInstanceOf(DerError);
  const failure = error as DerError;
  return { outcome: "error", error_id: failure.kind, offset: failure.offset };
}

function runDecode(testCase: FixtureCase, input: Uint8Array, limits: DerLimits): unknown {
  try {
    if (testCase.operation === "decode-one") {
      const [element, remainder] = decodeOne(input, limits);
      const projection = elementResult(element, 0) as { remainder_offset: number };
      expect(projection.remainder_offset).toBe(input.length - remainder.length);
      return projection;
    }
    return elementResult(decodeExact(input, limits), 0);
  } catch (error: unknown) { return errorResult(error); }
}

function runCursor(testCase: FixtureCase, input: Uint8Array, limits: DerLimits): unknown {
  const cursor = new DerCursor(input, limits);
  const events: unknown[] = [];
  for (const action of testCase.actions!) {
    if (action === "finish") {
      try { cursor.finish(); events.push({ outcome: "finished" }); } catch (error: unknown) { events.push(errorResult(error)); }
    } else {
      const offset = input.length - cursor.remaining.length;
      try {
        const element = cursor.read();
        events.push(element === undefined ? { outcome: "end" } : elementResult(element, offset));
      } catch (error: unknown) { events.push(errorResult(error)); }
    }
  }
  return { events, elements_read: cursor.elementsRead, remaining_offset: input.length - cursor.remaining.length };
}

describe("DER TLV v1 portable conformance", () => {
  it("pins the closed profile", () => {
    expect(fixture.cases).toHaveLength(54);
    expect(fixture.error_ids).toHaveLength(17);
  });
  for (const testCase of fixture.cases) {
    it(testCase.id, () => {
      const input = bytes(testCase.input);
      const result = testCase.operation === "cursor" ? runCursor(testCase, input, configured(testCase)) : runDecode(testCase, input, configured(testCase));
      expect(result).toEqual(testCase.expected);
      if (testCase.redacted_input_hex !== undefined) expect(JSON.stringify(result)).not.toContain(testCase.redacted_input_hex);
    });
  }
  it("returns views over the caller buffer", () => {
    const input = Uint8Array.of(0x04, 0x01, 0x2a);
    const element = decodeExact(input, defaultLimits());
    input[2] = 0x7f;
    expect([...element.value]).toEqual([0x7f]);
  });
});
